src/time_stream/aggregation.py

```diff
@@ -1,573 +1,354 @@
+        expected_count_name = f"expected_count_{ts.time_name}"
+        # For some aggregations, the expected count is a constant so use that if possible.
+        # For example, when aggregating 15-minute data over a day, the expected count is always 96.
+        count = ts.periodicity.count(period)
+        if count > 0:
+            return pl.lit(count).alias(expected_count_name)
 
-        if len(missing_criteria) != 1:
-            raise ValueError(f"missing_criteria argument should contain only one key, not {len(missing_criteria)}")
+        # Variable length periods need dynamic calculation, based on the start and end of a period interval
+        start_expr = pl.col(ts.time_name)
+        end_expr = pl.col(ts.time_name).dt.offset_by(period.pl_interval)
 
-        supplied_key = list(missing_criteria.keys())[0]
+        # This contains 2 cases:
+        if ts.periodicity.timedelta:
+            # 1. If the data we are aggregating is not monthly, then each interval we are aggregating has
+            #    a constant length, so (end - start) / interval will be the expected count.
+            micros = ts.periodicity.timedelta // timedelta(microseconds=1)
+            return ((end_expr - start_expr).dt.total_microseconds() // micros).alias(expected_count_name)
 
-        if supplied_key not in MissingCriteriaOptions:
-            raise KeyError(
-                f"missing_criteria option should be one of"
-                f"{[options.value for options in list(MissingCriteriaOptions)]} "
-                f"not '{supplied_key}'"
+        else:
+            # 2. If the data we are aggregating is month-based, then there is no simple way to do the calculation,
+            #    so use Polars to create a Series of date ranges lists and get the length of each list.
+            #
+            #    Note: This method will work for above use cases also, but if periodicity is small and the aggregation
+            #    period is large, it consumes too much memory and causes performance problems. For example, aggregating
+            #    1 microsecond data over a calendar year involves the creation length
+            #    1000_000 * 60 * 60 * 24 * 365 arrays which will probably fail with an out-of-memory error.
+            return (
+                pl.datetime_ranges(start_expr, end_expr, interval=ts.periodicity.pl_interval, closed="right")
+                .list.len()
+                .alias(expected_count_name)
             )
 
-        return {"method": f"_{supplied_key}", "limit": missing_criteria[supplied_key]}
-
-    def validate_aggregation(
-        self, column_name: str, missing_criteria: Mapping[str, Union[int | float]]
-    ) -> pl.DataFrame:
-        """Check the aggregated dataframe satisfies missing value criteria.
+    @staticmethod
+    def _missing_data_expr(
+        ts: TimeSeries, columns: List[str], missing_criteria: Optional[Tuple[str, Union[float, int]]] = None
```

Comment:
`missing_criteria` is a different type here than that passed in the base module. Need to align. Not fussed if `tuple` or `map`